pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/groupby/DefaultGroupByExecutor.java

```diff
@@ -183,4 +194,26 @@ public GroupKeyGenerator getGroupKeyGenerator() {
   public GroupByResultHolder[] getGroupByResultHolders() {
     return _groupByResultHolders;
   }
+
+  private Integer getGroupByResultHolderCapacityBasedOnFilterPredicate(QueryContext queryContext) {
+    if (queryContext.getFilter() == null || queryContext.getGroupByExpressions() == null) {
```

Comment:
Unless there are groupByExpressions, we wouldn't come in to this class. It's better for this to be an assert if at all you want to ensure this case. 

I think a better check here is to see if the groupByExpression contains only a single column (assuming that is the case you are optimizing)

Reply:
Refactor the logic to handle both single and multi groupby